### PR TITLE
[LLM Runtime] Beam Search Support of Fused Attention

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: Feature
API not changed

## Description
As title.

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR

### MLPerf accuracy
q4_j1_bf16_pc: {'rouge1': 42.8563, 'rouge2': 19.8546, 'rougeL': 29.4987, 'rougeLsum': 40.0661, 'gen_len': 4581237, 'gen_num': 13368}
> A bit low, but probably due to the calculation precision itself

### Local Run
<details>
<summary>`python -m unittest tests/test_llm_runtime.py -vk beam` with single prompt:</summary>

```diff
diff --git a/intel_extension_for_transformers/llm/runtime/graph/tests/test_llm_runtime.py b/intel_extension_for_transformers/llm/runtime/graph/tests/test_llm_runtime.py
index 5c93edaa07..0bc2392534 100644
--- a/intel_extension_for_transformers/llm/runtime/graph/tests/test_llm_runtime.py
+++ b/intel_extension_for_transformers/llm/runtime/graph/tests/test_llm_runtime.py
@@ -49,7 +49,7 @@ class TestLLMRUNTIME(unittest.TestCase):
 
         tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
         inputs = tokenizer(prompt, return_tensors="pt")
-        
+
         pt_logits = torch.load("/tf_dataset2/inc-ut/nlptoolkit_ut_model/llama2_pt_logits.pth")[:,-1]
         pt_generate_ids = torch.load("/tf_dataset2/inc-ut/nlptoolkit_ut_model/llama2_pt_generate_ids.pth")[0].tolist()
         print(tokenizer.decode(pt_generate_ids))
@@ -70,22 +70,22 @@ class TestLLMRUNTIME(unittest.TestCase):
             # "jblas_int8": WeightOnlyQuantConfig(compute_dtype="bf16", weight_dtype="int8", use_cache=True),
             }
         for config_type in woq_configs:
-            itrex_model = AutoModel.from_pretrained(model_name, quantization_config=woq_configs[config_type], 
+            itrex_model = AutoModel.from_pretrained(model_name, quantization_config=woq_configs[config_type],
                                                     use_llm_runtime=True, trust_remote_code=True)
             itrex_logits = itrex_model(inputs.input_ids)
             print(config_type, cmpData(pt_logits.detach().numpy().flatten(), itrex_logits.flatten()))
 
 
     def test_beam_search(self):
-        model_name = "/tf_dataset2/models/pytorch/gpt-j-6B"  # or local path to model
+        model_name = "/home/dingyi/gpt-j-6B"  # or local path to model
         prompts = [
            "she opened the door and see",
-           "tell me 10 things about jazz music",
-           "What is the meaning of life?",
-           "To be, or not to be, that is the question: Whether 'tis nobler in the mind to suffer"\
-            " The slings and arrows of outrageous fortune, "\
-            "Or to take arms against a sea of troubles."\
-            "And by opposing end them. To die—to sleep,"
+        #    "tell me 10 things about jazz music",
+        #    "What is the meaning of life?",
+        #    "To be, or not to be, that is the question: Whether 'tis nobler in the mind to suffer"\
+        #     " The slings and arrows of outrageous fortune, "\
+        #     "Or to take arms against a sea of troubles."\
+        #     "And by opposing end them. To die—to sleep,"
             ]
 
         tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True,
@@ -95,7 +95,10 @@ class TestLLMRUNTIME(unittest.TestCase):
         inputs = tokenizer(prompts, padding=True, return_tensors='pt')
 
         # pytorch fp32
-        pt_generate_ids = torch.load("/tf_dataset2/inc-ut/nlptoolkit_ut_model/beam_pt_generate_ids.pth").tolist()
+        pt_model = AutoModelForCausalLM.from_pretrained(model_name, trust_remote_code=True)
+        pt_model.eval()
+        pt_generate_ids = pt_model.generate(**inputs, max_new_tokens=128, min_new_tokens=30,
+                                            early_stopping=True, num_beams=4).tolist()
 
         # llm runtime fp32
         woq_config = WeightOnlyQuantConfig(not_quant=True, use_cache=True)
```
</details>

```
FP32 model will be used.
ARCH_REQ_XCOMP_PERM XTILE_DATA successful.
AVX:1 AVX2:1 AVX512F:1 AVX_VNNI:1 AVX512_VNNI:1 AMX_INT8:1 AMX_BF16:1 AVX512_BF16:1 AVX512_FP16:1
beam_size: 4, do_sample: 0, top_k: 40, top_p: 0.950000
model.cpp: loading model from runtime_outs/ne_gptj_f32.bin
init: n_vocab    = 50400
init: n_embd     = 4096
init: n_mult     = 256
init: n_head     = 16
init: n_layer    = 28
init: n_rot      = 64
init: n_ff       = 16384
init: n_parts    = 1
load: ne ctx size = 23082.36 MB
load: mem required  = 31274.36 MB (+ memory per state)
............................................................................................
model_init_from_file: support_jblas_kv = 1
model_init_from_file: kv self size =  966.00 MB
ok

----------------------------------------------------------------------
Ran 1 test in 632.999s

OK
```

## Dependency Change?
No

### CODE REVIEW

**Review by zhentaoyu — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 1]
```diff
@@ -2052,6 +2052,78 @@ void jblas_reordered_attn_fp32_shift_rope_k(char* cache, const ne_fp16_t* cossin
     }
 }
 
+template <bool zero_padding>
+void jblas_fusion_attn_fp32_batch_cpy_k_(const jblas_fusion_attn_fp32_batch_cpy_kv_args_t* params) {
+  const auto p = *params;
+  const auto pad_headsize = padto(p.head_size, 32);
+  const auto pad_seq_max = padto(p.seq_max, 48);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2059]: I think 48 and 32 are from Bestla kernels, make it meaningful
```
const static N_PAD=48;
const auto pad_seq_max = padto(p.seq_max, N_PAD,)
```

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 2]
```diff
@@ -2052,6 +2052,78 @@ void jblas_reordered_attn_fp32_shift_rope_k(char* cache, const ne_fp16_t* cossin
...
+  for (int ihn = 0; ihn < p.heads_kv; ++ihn) {
+    const auto dst = reinterpret_cast<bf16*>(p.dst) + ihn * step_head_num;
+    const auto src = reinterpret_cast<bf16*>(p.src) + ihn * step_head_num;
+
+    if (size_unaligned_cpy) {
+      for (int j = 0; j < pad_headsize; j += 2) {  // K-dim padding for QK_GEMM
+        const auto ii = p.seq_off % 48;
+        const auto i_blk = p.seq_off - ii;
+        const auto off = i_blk * pad_headsize + ii * 2 + j * 48;
+        memcpy(dst + off, src + off, sizeof(bf16) * 2 * size_unaligned_cpy);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2074]: seems for ascending ```j```, src and dst are contiguous, then why a loop here?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 3]
```diff
@@ -2052,6 +2052,78 @@ void jblas_reordered_attn_fp32_shift_rope_k(char* cache, const ne_fp16_t* cossin
...
+  const auto seq_end_aligned = padto(p.seq_off + p.seq_size, 32);
+  const auto seq_size_aligned = seq_end_aligned - seq_off_aligned;
+#pragma omp parallel for collapse(2)
+  for (int ihn = 0; ihn < p.heads_kv; ++ihn) {
+    for (int j = 0; j < p.head_size; j += 48) {
+      const auto dst = reinterpret_cast<bf16*>(p.dst) + ihn * step_head_num + pad_seq_max * j;
+      const auto src = reinterpret_cast<bf16*>(p.src) + ihn * step_head_num + pad_seq_max * j;
+      if (p.seq_off != seq_off_aligned) {
+        const auto off = (seq_off_aligned - 2) * 48 + 1;
+        for (int jj = 0; jj < 48; ++jj) src[off + jj * 2] = dst[off + jj * 2];
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: I think this will greatly hurt performance, any better solution?

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: maybe mask load is better since 48 is static

**Review by airMeng — CHANGES_REQUESTED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 2 above)
Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2074]: my fault

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 3 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: > I think this will greatly hurt performance, any better solution?

Talking about performance, we need a way to profile non-eval cost.

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: I think the cost is neglectable and it don't worth the effort to optimized it. There are only (head_size elements) process in this way vs (seq_size*head_size elements) processed in `memcpy`.

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: But it seems to be a bug here! I should copy src to dst

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: > > maybe mask load is better since 48 is static
> 
> I think the cost is neglectable and it don't worth the effort to optimized it. There are only (head_size elements, note the outer loop is in step of 48) process in this way vs (seq_size*head_size elements) processed in `memcpy`.

@yuchengliu1 can you provide a unified vectorized function to simplify these effort?

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: > But it seems to be a bug here! I should copy src to dst

then why the output are correct? or they not?

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: Maybe only errors of one token don't affect that much

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:2109]: For anyone who ran into here, you can manipulate compiler carefully https://godbolt.org/z/o6EEz9x44

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp]: Updated in 92ff74ad17680cee6a0977f1626d86eea7bfad03

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 3 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp]: > But it seems to be a bug here! I should copy src to dst

Fixed in 92ff74ad17680cee6a0977f1626d86eea7bfad03

**Comment by DDEle** [comment]
TODO: re-run MLPerf accuracy

**Review by airMeng — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp [hunk 4]
```diff
@@ -1561,7 +1561,7 @@ struct model_context* model_init_from_gpt_params(const gpt_params& params) {
   // printf("n_head_kv=%s,multi_query_group_num=%s",model_hparams.n_head_kv,model_hparams.multi_query_group_num);
   NE_ASSERT(("Can not set n_head_kv and multi_query_group_num at the same time",
              model_hparams.n_head_kv == 0 || model_hparams.multi_query_group_num == 0 ||
-                 model_hparams.n_head_kv != model_hparams.multi_query_group_num));
+                 model_hparams.n_head_kv == model_hparams.multi_query_group_num));
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_utils.cpp:1564]: Fixed bug
